File: frontend/src/utils/energyRecommendation.ts
import type { DayPrediction } from "./mlPredictor";

export type EnergySource = "Solar" | "Wind" | "Hydro" | "Thermal" | "Nuclear";

export interface DayEnergyRecommendation {
  date: string;
  best: EnergySource;
  scores: Record<EnergySource, number>;
  reason: string;
}

export interface EnergyRecommendationResult {
  overallBest: EnergySource;
  overallReason: string;
  daily: DayEnergyRecommendation[];
}

function clamp(val: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, val));
}

function scoreSolar(p: DayPrediction): number {
  let score = 50;
  if (p.weather_condition === "sunny") score += 35;
  else if (p.weather_condition === "cloudy") score -= 15;
  else if (p.weather_condition === "rainy" || p.weather_condition === "stormy")
    score -= 30;
  if (p.temperature >= 25 && p.temperature <= 40) score += 10;
  else if (p.temperature < 20) score -= 10;
  if (p.humidity < 40) score += 5;
  else if (p.humidity >= 60) score -= 10;
  else if (p.humidity >= 80) score -= 20;
  return clamp(score, 0, 100);
}

function scoreWind(p: DayPrediction): number {
  let score = 30;
  score += clamp(p.wind_speed * 2.5, 0, 55);
  if (p.weather_condition === "windy") score += 15;
  else if (p.weather_condition === "stormy") score += 10;
  return clamp(score, 0, 100);
}

function scoreHydro(p: DayPrediction): number {
  let score = 40;
  if (p.weather_condition === "rainy") score += 35;
  else if (p.weather_condition === "stormy") score += 20;
  if (p.humidity > 80) score += 15;
  else if (p.humidity > 70) score += 8;
  return clamp(score, 0, 100);
}

function scoreThermal(p: DayPrediction): number {
  let score = 55;
  if (p.demand_mw > 4000) score += 20;
  else if (p.demand_mw > 3000) score += 10;
  return clamp(score, 0, 100);
}

function scoreNuclear(p: DayPrediction): number {
  let score = 50;
  if (p.demand_mw > 4000) score += 15;
  else if (p.demand_mw > 3000) score += 8;
  return clamp(score, 0, 100);
}

const REASON_MAP: Record<EnergySource, string> = {
  Solar: "Clear skies and optimal temperatures favor solar generation",
  Wind: "High wind speeds make wind turbines most efficient",
  Hydro: "High humidity and rainfall indicators support hydro generation",
  Thermal: "High demand periods call for reliable thermal baseload",
  Nuclear: "Steady baseload from nuclear suits high-demand conditions",
};

export function recommendEnergySource(
  predictions: DayPrediction[],
): EnergyRecommendationResult {
  const daily: DayEnergyRecommendation[] = predictions.map((p) => {
    const scores: Record<EnergySource, number> = {
      Solar: scoreSolar(p),
      Wind: scoreWind(p),
      Hydro: scoreHydro(p),
      Thermal: scoreThermal(p),
      Nuclear: scoreNuclear(p),
    };
    const best = (Object.keys(scores) as EnergySource[]).reduce((a, b) =>
      scores[a] >= scores[b] ? a : b,
    );
    return { date: p.date, best, scores, reason: REASON_MAP[best] };
  });

  const tally: Record<EnergySource, number> = {
    Solar: 0,
    Wind: 0,
    Hydro: 0,
    Thermal: 0,
    Nuclear: 0,
  };
  for (const d of daily) tally[d.best]++;
  const overallBest = (Object.keys(tally) as EnergySource[]).reduce((a, b) =>
    tally[a] >= tally[b] ? a : b,
  );

  return { overallBest, overallReason: REASON_MAP[overallBest], daily };
}

File: backend_python/utils/energy_recommendation.py
from typing import List, Dict

EnergySource = str

REASON_MAP = {
    "Solar": "Clear skies and optimal temperatures favor solar generation",
    "Wind": "High wind speeds make wind turbines most efficient",
    "Hydro": "High humidity and rainfall indicators support hydro generation",
    "Thermal": "High demand periods call for reliable thermal baseload",
    "Nuclear": "Steady baseload from nuclear suits high-demand conditions",
}


def clamp(val: float, min_val: float, max_val: float) -> float:
    return max(min_val, min(max_val, val))


def score_solar(p: Dict) -> float:
    score = 50
    if p.get("weather_condition") == "sunny":
        score += 35
    elif p.get("weather_condition") == "cloudy":
        score -= 15
    elif p.get("weather_condition") in ["rainy", "stormy"]:
        score -= 30

    temp = p.get("temperature", 25)
    if 25 <= temp <= 40:
        score += 10
    elif temp < 20:
        score -= 10

    humid = p.get("humidity", 60)
    if humid < 40:
        score += 5
    elif humid >= 60:
        score -= 10
    elif humid >= 80:
        score -= 20

    return clamp(score, 0, 100)


def score_wind(p: Dict) -> float:
    score = 30
    wind_speed = p.get("wind_speed", 0)
    score += clamp(wind_speed * 2.5, 0, 55)

    weather = p.get("weather_condition", "")
    if weather == "windy":
        score += 15
    elif weather == "stormy":
        score += 10

    return clamp(score, 0, 100)


def score_hydro(p: Dict) -> float:
    score = 40
    weather = p.get("weather_condition", "")
    if weather == "rainy":
        score += 35
    elif weather == "stormy":
        score += 20

    humid = p.get("humidity", 0)
    if humid > 80:
        score += 15
    elif humid > 70:
        score += 8

    return clamp(score, 0, 100)


def score_thermal(p: Dict) -> float:
    score = 55
    demand = p.get("demand_mw", 0)
    if demand > 4000:
        score += 20
    elif demand > 3000:
        score += 10
    return clamp(score, 0, 100)


def score_nuclear(p: Dict) -> float:
    score = 50
    demand = p.get("demand_mw", 0)
    if demand > 4000:
        score += 15
    elif demand > 3000:
        score += 8
    return clamp(score, 0, 100)


def recommend_energy_source(predictions: List[Dict]) -> Dict:
    energy_sources = ["Solar", "Wind", "Hydro", "Thermal", "Nuclear"]

    daily = []
    for p in predictions:
        scores = {
            "Solar": score_solar(p),
            "Wind": score_wind(p),
            "Hydro": score_hydro(p),
            "Thermal": score_thermal(p),
            "Nuclear": score_nuclear(p),
        }

        best = max(scores, key=scores.get)

        daily.append(
            {
                "date": p["date"],
                "best": best,
                "scores": {k: round(v) for k, v in scores.items()},
                "reason": REASON_MAP[best],
            }
        )

    tally = {src: 0 for src in energy_sources}
    for d in daily:
        tally[d["best"]] += 1

    overall_best = max(tally, key=tally.get)

    return {
        "overallBest": overall_best,
        "overallReason": REASON_MAP[overall_best],
        "daily": daily,
    }
